### PR TITLE
[M1166] Add ACA Covariates to sites

### PR DESCRIPTION
**Pull request by gridcell**
- new model `Covariate`
- DJ command to backfill `update_covariates`
- Site pre_save signal to fetch and save covariates

### CODE REVIEW

**Comment by gridcell** [comment]
[include Coral Atlas covariates in Collect field reports and api obs/transect/site endpoints](https://trello.com/c/WwmQ01H1/1166-include-coral-atlas-covariates-in-collect-field-reports-and-api-obs-transect-site-endpoints)

File: src/api/covariates/coral_atlas.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            class_type = (map_asset.get("class_type") or "").lower()
+            if not class_type or class_type not in (
+                self.BENTHIC_CLASS_TYPE,
+                self.GEOMORPHIC_CLASS_TYPE,
+            ):
+                continue
+            classes = self._parse_classes(map_asset.get("classes"))
+            output[f"aca_{class_type}"].extend(classes)
+
+        return dict(
```
Comment (saanobhaai) [inline src/api/covariates/coral_atlas.py:54]: Shouldn't this be a list of dicts?

File: src/api/covariates/__init__.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    if not results:
+        return
+
+    result = results[0]
+
+    data_date = result["date"]
+    requested_date = result["requested_date"]
+    aca_covariates = result.get("covariates") or dict()
+    aca_benthic = aca_covariates["aca_benthic"]
```
Comment (saanobhaai) [inline src/api/covariates/__init__.py:31]: protect against potential `KeyError` here?

File: src/api/coral_atlas.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    return CoralAtlasDataPoint(
+        feature.get("lng"),
+        feature.get("lat"),
+        feature.get("benthic"),
+        feature.get("geomorphic"),
+        feature.get("reeftype"),
+    )
+
+
+def get_data_points(points):
```
Comment (saanobhaai) [inline src/api/coral_atlas.py:39]: Hang on, is this file used?

File: src/api/models/mermaid.py [hunk 4]
```diff
@@ -1567,3 +1567,27 @@ class ArchivedRecord(models.Model):
...
+        "aca_benthic",
+        "aca_geomorphic",
+    )
+
+    site = models.ForeignKey(
+        "Site",
+        related_name="covariates",
+        on_delete=models.CASCADE
+    )
+    name = models.CharField(max_length=100)
```
Comment (saanobhaai) [inline src/api/models/mermaid.py:1583]: `choices=SUPPORTED_COVARIATES`?

File: src/api/covariates/coral_atlas.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                    )
+                )
+
+            for future in futures:
+                response.append(future.result())
+
+        return response
+
+
+if __name__ == "__main__":
```
Comment (saanobhaai) [inline src/api/covariates/coral_atlas.py:83]: I guess ideally this would get implemented as a test?

**Review by saanobhaai — CHANGES_REQUESTED** [review]
Masterful -- probably don't need to actually change much to address comments here

File: src/api/covariates/coral_atlas.py (hunk 1 above)
Comment (gridcell) [reply src/api/covariates/coral_atlas.py:54]: this only deals with one location and all the ACA covariates at once

File: src/api/reports/formatters.py [hunk 6]
```diff
@@ -118,3 +118,25 @@ def to_percent_cover(value, field, row, serializer_instance):
 
     return f"Average quadrat count: {quadrat_count_avg}, Average hard coral %: {percent_hard_avg_avg}, " \
            f"Average soft coral %: {percent_soft_avg_avg}, Average macroalgae %: {percent_algae_avg_avg}"
+
+
+def _to_aca_covariate(value, covariate_key):
```
[the remark this answers is not in the document]
Comment (gridcell) [reply src/api/reports/formatters.py:123]: Going to keep it with the same name, not sure if other covariates will have the same `dict` structure, can refactor then

**Pull request merged**